syntax = "proto3";

package driver;
import "task.proto";
import "worker.proto";



message AssignTaskRequest {
  worker.Worker worker = 1;
}

message AssignTaskResponse {
  optional task.Task task = 1;
  optional string error = 2; 
}

message UpdateTaskStatusRequest {
  worker.Worker worker = 1;
  task.Task task = 2;
}

message UpdateTaskStatusResponse {
  bool ack = 1;
}


message CompleteTaskRequest {
  task.Task task = 1;
  string worker_id = 2;
}
message CompleteTaskResponse {
  bool ack = 1;
}


// Servizio fornito dal driver per gestire i task MapReduce
service MapReduceDriver {
  // Metodo per assegnare un task a un worker
  rpc AssignTask(AssignTaskRequest) returns (AssignTaskResponse);

  //update task status
  rpc UpdateTaskStatus(UpdateTaskStatusRequest) returns (UpdateTaskStatusResponse);

  // Segnala il completamento di un task da parte di un worker
  rpc CompleteTask(CompleteTaskRequest) returns (CompleteTaskResponse);
  
}